src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/MemberMatch/MemberMatchService.cs

```diff
@@ -0,0 +1,174 @@
+        {
+            EnsureArg.IsNotNull(searchServiceFactory, nameof(searchServiceFactory));
+            EnsureArg.IsNotNull(resourceDeserializer, nameof(resourceDeserializer));
+            EnsureArg.IsNotNull(searchIndexer, nameof(searchIndexer));
+            EnsureArg.IsNotNull(searchParameterDefinitionManagerResolver, nameof(searchParameterDefinitionManagerResolver));
+            EnsureArg.IsNotNull(expressionParser, nameof(expressionParser));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _searchServiceFactory = searchServiceFactory;
+            _resourceDeserializer = resourceDeserializer;
+            _searchIndexer = searchIndexer;
+            _expressionParser = expressionParser;
+            _coverageBeneficiaryParameter = searchParameterDefinitionManagerResolver().GetSearchParameter("Coverage", "beneficiary");
+            _logger = logger;
+        }
+
+        public async Task<ResourceElement> FindMatch(ResourceElement coverage, ResourceElement patient, CancellationToken cancellationToken)
+        {
+            var searchOptions = new SearchOptions();
+            searchOptions.MaxItemCount = 2;
+            searchOptions.Sort = new List<(SearchParameterInfo, SortOrder)>();
+            searchOptions.UnsupportedSearchParams = new List<Tuple<string, string>>();
+            searchOptions.Expression = CreateSearchExpression(coverage, patient);
+
+            SearchResult results = null;
+            try
+            {
+                using IScoped<ISearchService> search = _searchServiceFactory();
+                results = await search.Value.SearchAsync(searchOptions, cancellationToken);
+            }
+
+            // Generic catch for any exception happened during search.
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Generic problem in MemberMatch service.");
+                throw new MemberMatchMatchingException(Core.Resources.MemberMatchNoMatchFound);
+            }
+
+            return CreatePatientWithIdentity(patient, results);
+        }
+
+        private ResourceElement CreatePatientWithIdentity(ResourceElement patient, SearchResult results)
+        {
+            var searchMatchOnly = results.Results.Where(x => x.SearchEntryMode == ValueSets.SearchEntryMode.Match).ToList();
+            if (searchMatchOnly.Count > 1)
+            {
+                throw new MemberMatchMatchingException(Core.Resources.MemberMatchMultipleMatchesFound);
+            }
+
+            if (searchMatchOnly.Count == 0)
+            {
+                throw new MemberMatchMatchingException(Core.Resources.MemberMatchNoMatchFound);
+            }
+
+            var match = searchMatchOnly[0];
+            var element = _resourceDeserializer.Deserialize(match.Resource);
+            var foundPatient = element.ToPoco<Patient>();
+            var id = foundPatient.Identifier.Where(x => x.Type.Coding.Exists(x => x.Code == "MB")).FirstOrDefault();
```

Comment:
>x.Type.Coding.Exists [](start = 56, length = 20)

We will have to add a null check here since not all Identifiers might have Coding